File: service/app/decision_store.py
"""SQLite-backed Decision Record persistence.

Written in phases (pending -> processing -> completed/gitea_call_failed) so
a repeated POST of the same Raw Report can check for a prior record by
report hash before doing any LLM or Gitea work.
"""

from __future__ import annotations

import sqlite3
import threading

from app.schemas import DecisionRecord

SCHEMA = """
CREATE TABLE IF NOT EXISTS decision_records (
    report_hash TEXT PRIMARY KEY,
    payload TEXT NOT NULL
);
"""


class SqliteDecisionStore:
    """One long-lived connection guarded by a lock (sqlite3.Connection is not
    thread-safe on its own; check_same_thread=False plus the lock lets
    FastAPI's threadpool-run sync handlers share it safely).
    """

    def __init__(self, path: str):
        self._conn = sqlite3.connect(path, check_same_thread=False)
        self._lock = threading.Lock()
        with self._lock:
            self._conn.execute(SCHEMA)
            self._conn.commit()

    def save(self, record: DecisionRecord) -> None:
        with self._lock:
            self._conn.execute(
                "INSERT INTO decision_records (report_hash, payload) VALUES (?, ?) "
                "ON CONFLICT(report_hash) DO UPDATE SET payload = excluded.payload",
                (record.report_hash, record.model_dump_json()),
            )
            self._conn.commit()

    def get(self, report_hash: str) -> DecisionRecord | None:
        with self._lock:
            row = self._conn.execute(
                "SELECT payload FROM decision_records WHERE report_hash = ?", (report_hash,)
            ).fetchone()
        if row is None:
            return None
        return DecisionRecord.model_validate_json(row[0])

File: service/src/decisions/decision-store.ts
/**
 * Postgres-backed (Prisma) Decision Record persistence (issue #59, replacing
 * the SQLite-backed implementation from before #58/#59). Mirrors
 * app/decision_store.py.
 *
 * Written in phases (pending -> processing -> completed/gitea_call_failed) so
 * a repeated POST of the same Raw Report can check for a prior record by
 * report hash before doing any LLM or Gitea work.
 */

import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { DecisionRecord } from '../reports/types';
import { fromDecisionRow, toDecisionCreateInput, toDecisionReplaceUpdateInput } from './decision-record.mapper';
import { PrismaService } from './prisma.service';

const INCLUDE_CHILDREN = {
  duplicateCandidates: true,
  tokenUsages: true,
  stageTimings: true,
} as const;

function isUniqueConstraintViolation(error: unknown): boolean {
  return error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2002';
}

@Injectable()
export class DecisionStore {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * Atomic insert-or-bail (F3 audit finding: processReport's prior
   * get-then-save had an await point between the read and the write, so two
   * concurrent identical POSTs could both observe no existing record and
   * both proceed to do the LLM/Gitea work). The `report_hash` unique
   * constraint makes this atomic at the database level regardless of
   * concurrency -- returns whether this call actually created the row.
   */
  async tryClaim(record: DecisionRecord): Promise<boolean> {
    try {
      await this.prisma.decision.create({ data: toDecisionCreateInput(record) });
      return true;
    } catch (error) {
      if (isUniqueConstraintViolation(error)) return false;
      throw error;
    }
  }

  async save(record: DecisionRecord): Promise<void> {
    await this.prisma.decision.upsert({
      where: { reportHash: record.report_hash },
      create: toDecisionCreateInput(record),
      update: toDecisionReplaceUpdateInput(record),
    });
  }

  async get(reportHash: string): Promise<DecisionRecord | null> {
    const row = await this.prisma.decision.findUnique({
      where: { reportHash },
      include: INCLUDE_CHILDREN,
    });
    return row ? fromDecisionRow(row) : null;
  }
}
